[AzureDNS] Add support for Managed Identity

jon-walton opened the pull request:
**What this PR does / why we need it**:

This adds support for azure managed identities. It means you no longer need to add a client ID in the issuer or create a secret.

**Special notes for your reviewer**:
I'm not entirely sure if I've done the ```// +optional``` comments correctly?

**Release note**:
```release-note
Add support for Azure Managed Identity
```

jon-walton commented:
ping @kragniz , anyone able to take a look at this?

donchev7 commented:
Hey @jon-walton I was thinking about a similar PR but instead of having an explicit bool **useManagedIdentity** why not use **clientSecretSecretRef** ? 

If its defined MSI shouldn't be used.

donchev7 commented:
@kragniz   any comments?

jon-walton commented:
@donchev7 that's cleaner for sure.

Unfortunately I won't be prioritizing any time on this as it's been impossible to get any sort of acknowledgement from the jetstack team

jon-walton commented:
rebased to master and incorporated @donchev7 feedback.

I'll be testing this in my clusters over the coming days (I'll need to do the 0.6 to 0.10 upgrade)

jon-walton commented:
this rebase is working as expected on my clusters.

/assign @kragniz

twendt commented:
I tested it in my environment and it works great.

jon-walton commented:
/remove-lifecycle rotten

jon-walton commented:
ping @munnerz @kragniz

munnerz commented on pkg/internal/apis/certmanager/validation/issuer.go:
```diff
@@ -264,9 +264,8 @@ func ValidateACMEIssuerDNS01Config(iss *v1alpha1.ACMEIssuerDNS01Config, fldPath
-				if len(p.AzureDNS.ClientID) == 0 {
-					el = append(el, field.Required(fldPath.Child("azuredns", "clientID"), ""))
+				if len(p.AzureDNS.ClientID) > 0 {
+					el = append(el, ValidateSecretKeySelector(&p.AzureDNS.ClientSecret, fldPath.Child("azuredns", "clientSecretSecretRef"))...)
```
I think we should amend this check to test:

* if the client ID is set, the clientSecret is required and must be valid
* if the clientSecret is set, clientID must be provided and must be valid
* if neither are set, nothing should be checked (i.e. ClientID is no longer required)

munnerz commented on pkg/apis/certmanager/v1alpha1/types_issuer.go:
```diff
@@ -504,8 +504,14 @@ type ACMEIssuerDNS01ProviderRoute53 struct {
+	// The ClientID of the service principal used to authenticate with the Azure API
+	// Not required if using Managed Identities
+	// +optional
 	ClientID string `json:"clientID"`
```
Add `,omitempty` to this

munnerz commented on pkg/apis/certmanager/v1alpha1/types_issuer.go:
```diff
@@ -504,8 +504,14 @@ type ACMEIssuerDNS01ProviderRoute53 struct {
+	// The secret key for the service principal.
+	// Required if ClientID is configured
+	// +optional
 	ClientSecret SecretKeySelector `json:"clientSecretSecretRef"`
```
I think this should be a pointer and have omitempty so we can detect nil values

munnerz commented on pkg/apis/certmanager/v1alpha1/types_issuer.go:
```diff
@@ -504,8 +504,14 @@ type ACMEIssuerDNS01ProviderRoute53 struct {
+	// +optional
 	ClientSecret SecretKeySelector `json:"clientSecretSecretRef"`
 
 	SubscriptionID string `json:"subscriptionID"`
```
Just confirming, this field is still required even when using managed identity? Also, if you are familiar with these options, it'd be awesome if you could add a comment to any fields missing one to help others in future 😄

jon-walton replied:
sure, i'll add comments explaining them :+1:

`ClientSecret` is not used when with managed identity

munnerz commented on pkg/issuer/acme/dns/azuredns/azuredns.go:
```diff
@@ -68,16 +68,31 @@ func NewDNSProviderCredentials(environment, clientID, clientSecret, subscription
+			return nil, err
+		}
+
+		token, err = adal.NewServicePrincipalTokenFromMSI(endpoint, env.ServiceManagementEndpoint)
```
I think we need to check if 'ambient credentials' are allowed to be used here too - you can pass it through from [here](https://github.com/jetstack/cert-manager/blob/cb4dbe0d94fce4dabb217b1f5c5e3706a4b9499c/pkg/issuer/acme/dns/dns.go#L193) to the provider constructor 😄

munnerz commented on docs/tasks/issuers/setup-acme/dns01/azuredns.rst:
```diff
@@ -53,8 +53,11 @@ You can configure the issuer like so:
+             # Requires aadPodIdentity to be configured in the cluster
              clientID: AZURE_SERVICE_PRINCIPAL_ID
              # A secretKeyRef to a service principal ClientSecret (password)
+             # Required if clientID is set
```
We have moved our documentation to github.com/cert-manager/website - would you mind opening a PR against [this](https://github.com/cert-manager/website/blob/master/content/en/docs/configuration/acme/dns01/azuredns.md) file with theese changes? 😄

munnerz replied:
It'd also be great if you could add a header/section specifically relating to using managed identity, i.e. `Authenticating using Azure Managed Identity` (with a brief blurb explaining why you'd do this, e.g. `In order to avoid managing a client ID and client secret for Azure, it is possible to authenticate to Azure using 'ambient credentials' which use [Azure Managed Identity](link here) to authenticate with the Azure API ...`)

munnerz commented on docs/tasks/issuers/setup-acme/dns01/azuredns.rst:
```diff
@@ -53,8 +53,11 @@ You can configure the issuer like so:
            azuredns:
              # Service principal clientId (also called appId)
+             # Optional. If omitted, a Managed Identity will be used for Azure API authentication.
+             # Requires aadPodIdentity to be configured in the cluster
```
Can we link to more information on setting up `aadPodIdentity` here?

jon-walton commented:
thanks @munnerz , I'll try get to this over the next week and will open a PR on the website repo

unguiculus commented:
I'd like this feature make it in soon. 

@jon-walton Could you move this forward? Otherwise, I'd be happy to look into it and to pick it up.

Bowbaq commented:
@jon-walton we are also interested in seeing this get in. Happy to assist if needed

jon-walton commented:
@munnerz apologies for the blackout; rebased with comments addressed. but I just saw #2681 , feel free to close this in favour of that pr
